compose/ui/ui/src/webCommonW3C/kotlin/androidx/compose/ui/platform/WebTextInputService.kt

```diff
@@ -28,43 +28,53 @@ import org.w3c.dom.events.KeyboardEvent
 internal interface InputAwareInputService {
     fun getOffset(rect: Rect): Offset
     fun processKeyboardEvent(keyboardEvent: KeyboardEvent)
-    fun isVirtualKeyboard(): Boolean
 }
 
 internal abstract class WebTextInputService : PlatformTextInputService, InputAwareInputService {
-    private val webImeInputService = WebImeInputService(this)
-    private val webKeyboardInputService = WebKeyboardInputService()
 
-    private fun delegatedService(): PlatformTextInputService {
-        return if (isVirtualKeyboard()) webImeInputService else webKeyboardInputService
-    }
+    private var backingTextArea: BackingTextArea? = null
+        set(value) {
+            field?.dispose()
+            field = value
+        }
 
     override fun startInput(
         value: TextFieldValue,
         imeOptions: ImeOptions,
         onEditCommand: (List<EditCommand>) -> Unit,
         onImeActionPerformed: (ImeAction) -> Unit
     ) {
-        delegatedService().startInput(value, imeOptions, onEditCommand, onImeActionPerformed)
+        backingTextArea =
+            BackingTextArea(
+                imeOptions = imeOptions,
+                onEditCommand = onEditCommand,
+                onImeActionPerformed = onImeActionPerformed,
+                processKeyboardEvent = this::processKeyboardEvent
+            )
+        backingTextArea?.register()
+
+        showSoftwareKeyboard()
     }
 
     override fun stopInput() {
-        delegatedService().stopInput()
+        backingTextArea?.dispose()
     }
 
     override fun showSoftwareKeyboard() {
-        delegatedService().showSoftwareKeyboard()
+        backingTextArea?.focus()
     }
 
     override fun hideSoftwareKeyboard() {
-        delegatedService().hideSoftwareKeyboard()
+        backingTextArea?.blur()
```

Comment:
I don't know the context but it seems strange (to blur instead of to hide keyboard)

Reply:
@terrakok this is the canonical way of doing so in general. 
It looks like hack, it feels like hack, it's a hack in some sense but the "official" one.

Thinks can be improved as soon as we'll decide to invoke experimental APIs like
https://developer.mozilla.org/en-US/docs/Web/API/VirtualKeyboard_API
but so far we try to avoid any experimental API not supported more or less universally.